src/analysis.ts

```diff
@@ -36,10 +39,55 @@ export class ValidationError extends Error {
 
 
 export class Analysis {
-  descriptors: DocumentDescriptor[];
+  private _descriptors: DocumentDescriptor[];
+  private _elementsByTagName = new Map<string, ElementDescriptor>();
+  private _elementsByPackageDir = new Map<string, ElementDescriptor[]>();
+  elementPaths = new Map<ElementDescriptor, string>();
+  locationOffsets = new Map<ElementDescriptor, LocationOffset>();
 
   constructor(descriptors: DocumentDescriptor[]) {
-    this.descriptors = descriptors;
+    this._descriptors = descriptors;
+    const packageGatherer = new PackageGatherer();
+    const elementsGatherer = new ElementGatherer();
+    new AnalysisWalker(this._descriptors).walk([
+      packageGatherer, elementsGatherer
+    ]);
+
+    const elements = elementsGatherer.elementDescriptors;
+
+    for (const element of elements) {
+      if (element.is) {
+        this._elementsByTagName.set(element.is);
+      }
+      let elementPath = elementsGatherer.elementPaths.get(element);
+      const matchingPackageDirs =
+          Array.from(packageGatherer.packageDirs)
+              .filter(dir => trimLeft(elementPath, '/').startsWith(dir));
+      const longestMatchingPackageDir =
+          matchingPackageDirs.sort((a, b) => b.length - a.length)[0] || '';
```

Comment:
actually, matching then getting the longest could be combined into one loop and it might be easier to follow